Fix filtering parts with indexHint for non analyzer (resubmit)

azat opened the pull request:
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix filtering parts with indexHint for non analyzer

Fixes: #54185 (cc @KochetovNicolai )
Resubmits: #54449 (cc @Avogar ) - it had been reverted in #54806, due to conflicts with newly added tests in #54563
Fixes: #54563 (Cc @rschu1ze) - add proper type check into virtual column utils

azat committed:
Fix filtering parts with indexHint for non analyzer

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>
(cherry picked from commit ffa82e9297c1d01f77793cc7b43aa4fb7bbec9c4)

robot-ch-test-poll1 commented:
<!-- automatic status comment for PR #54825 from azat/ClickHouse:fix-virtual-columns-filtering -->
*This is an automated comment for commit f3d7f6b6c6f215c23e5a87721a59258afd641ad3 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/54825/f3d7f6b6c6f215c23e5a87721a59258afd641ad3/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>✅ success</td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>⏳ pending</td></tr>
<tbody>
</table>

azat committed:
Check type of the filter expressions while filtering by virtual columns

This should fix filtering by virtual columns with non-uint8 types, i.e.
queries like:

    SELECT * FROM data WHERE 1.0

Fixes: 02346_full_text_search
Fixes: 00990_hasToken_and_tokenbf
v2: move check out from is_const to filterBlockWithQuery(), since in is_const there is no way to validate sets.
Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat commented:
>Stateless tests (ubsan) [1/2] — fail: 1, passed: 2784, skipped: 23

- `00002_log_and_exception_messages_formatting` - can be flaky (anyway this codes does not modified any exceptions)

>Integration tests (release) [3/4] — fail: 1, passed: 603

- `test_global_overcommit_tracker` - can be flaky
